File: internal/tts/mock.go
package tts

import (
	"context"
	"encoding/binary"
	"fmt"
	"math"
	"sync"
	"time"
)

// Mock is a deterministic Synthesizer for tests and for running the scheduler
// end to end without a GPU.
//
// It produces real WAV bytes — silence shaped like speech — so that anything
// downstream which decodes audio is genuinely exercised.
type Mock struct {
	// Latency is how long each call takes, simulating GPU time.
	Latency time.Duration
	// FailOn makes the given segment texts fail, for testing the fallback
	// paths. The value is the error returned.
	FailOn map[string]error
	// SecondsPerChar controls generated duration.
	SecondsPerChar float64

	mu    sync.Mutex
	calls []Request
}

func (m *Mock) Name() string { return "mock" }

func (m *Mock) Synthesize(ctx context.Context, req Request) (*Result, error) {
	if err := req.validate(); err != nil {
		return nil, err
	}
	if err := CheckLanguage(req.Language); err != nil {
		return nil, err
	}

	m.mu.Lock()
	m.calls = append(m.calls, req)
	m.mu.Unlock()

	if m.Latency > 0 {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(m.Latency):
		}
	}
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err, bad := m.FailOn[req.Text]; bad {
		return nil, err
	}

	spc := m.SecondsPerChar
	if spc == 0 {
		spc = 0.12 // roughly a natural reading pace for CJK
	}
	dur := time.Duration(float64(len([]rune(req.Text))) * spc * float64(time.Second))
	const rate = 24000
	return &Result{
		Audio:         toneWAV(rate, dur),
		SampleRate:    rate,
		Duration:      dur,
		Engine:        m.Name(),
		EngineVersion: "mock-1",
	}, nil
}

// Calls returns a copy of the requests seen so far.
func (m *Mock) Calls() []Request {
	m.mu.Lock()
	defer m.mu.Unlock()
	return append([]Request(nil), m.calls...)
}

// Reset clears recorded calls.
func (m *Mock) Reset() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.calls = nil
}

// toneWAV builds a 16-bit mono WAV of the given duration holding a quiet tone.
func toneWAV(rate int, d time.Duration) []byte {
	n := int(d.Seconds() * float64(rate))
	if n < 1 {
		n = 1
	}
	var b []byte
	put32 := func(v uint32) { b = binary.LittleEndian.AppendUint32(b, v) }
	put16 := func(v uint16) { b = binary.LittleEndian.AppendUint16(b, v) }

	b = append(b, "RIFF"...)
	put32(uint32(36 + n*2))
	b = append(b, "WAVEfmt "...)
	put32(16)
	put16(1)
	put16(1)
	put32(uint32(rate))
	put32(uint32(rate * 2))
	put16(2)
	put16(16)
	b = append(b, "data"...)
	put32(uint32(n * 2))
	for i := range n {
		v := 0.1 * math.Sin(2*math.Pi*220*float64(i)/float64(rate))
		b = binary.LittleEndian.AppendUint16(b, uint16(int16(v*32767)))
	}
	return b
}

// ErrSynthetic is a convenience for tests that need a retryable failure.
var ErrSynthetic = fmt.Errorf("tts: synthetic failure")

File: server/src/main/kotlin/mimimoto/tts/Mock.kt
package mimimoto.tts

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.time.Duration
import java.util.Collections
import kotlin.math.PI
import kotlin.math.roundToInt
import kotlin.math.sin

/**
 * A deterministic [Synthesizer] for tests and for running the scheduler end to
 * end without a GPU.
 *
 * It produces real WAV bytes — a quiet tone shaped like a segment — so anything
 * downstream that decodes audio is genuinely exercised.
 */
class MockSynthesizer(
    /** Simulated GPU time per call. */
    var latency: Duration = Duration.ZERO,
    /** Texts that should fail, and the exception to raise for each. */
    val failOn: MutableMap<String, Exception> = mutableMapOf(),
    /** Controls generated duration; the default is a natural CJK reading pace. */
    val secondsPerChar: Double = 0.12,
) : Synthesizer {

    override val name = "mock"

    /**
     * Matches the real engine's coverage rather than accepting everything, so a
     * test that passes here would also pass in production. A mock that is more
     * permissive than the thing it stands in for hides exactly the failures it
     * is there to catch.
     */
    override val languages: Set<String> = setOf("zh", "yue", "en", "ja", "ko")

    private val recorded = Collections.synchronizedList(mutableListOf<Request>())

    /** A snapshot of the requests seen so far. */
    val calls: List<Request> get() = synchronized(recorded) { recorded.toList() }

    fun reset() = synchronized(recorded) { recorded.clear() }

    override fun synthesize(request: Request): Result {
        request.validate()
        primaryLanguage(request.language, languages) // throws for an unsupported tag
        recorded += request

        if (!latency.isZero) Thread.sleep(latency.toMillis())
        failOn[request.text]?.let { throw it }

        val duration = Duration.ofMillis(
            (request.text.codePointCount(0, request.text.length) * secondsPerChar * 1000).toLong()
        )
        val rate = 24_000
        return Result(
            audio = toneWav(rate, duration),
            sampleRate = rate,
            duration = duration,
            engine = name,
            engineVersion = "mock-1",
        )
    }
}

/** A 16-bit mono WAV of the given duration holding a quiet tone. */
internal fun toneWav(rate: Int, duration: Duration): ByteArray {
    val n = maxOf((duration.toMillis() * rate / 1000).toInt(), 1)
    val buf = ByteBuffer.allocate(44 + n * 2).order(ByteOrder.LITTLE_ENDIAN)
    buf.put("RIFF".toByteArray())
    buf.putInt(36 + n * 2)
    buf.put("WAVEfmt ".toByteArray())
    buf.putInt(16)
    buf.putShort(1)
    buf.putShort(1)
    buf.putInt(rate)
    buf.putInt(rate * 2)
    buf.putShort(2)
    buf.putShort(16)
    buf.put("data".toByteArray())
    buf.putInt(n * 2)
    for (i in 0 until n) {
        val v = 0.1 * sin(2 * PI * 220 * i / rate)
        buf.putShort((v * 32767).roundToInt().toShort())
    }
    return buf.array()
}
